### PR TITLE
Move Parallelism usage from Apex -> Megatron Core

### PR DESCRIPTION
**Pull request by aklife97**
# What does this PR do ?

This PR moves model parallelism in NeMo to use Megatron-core instead of Apex.

* All models use dataloader_iter in train and eval step (except eval for T5 p-tuning)
* All pre-training models that use a constant sequence length sample based on microbatches
* All downstream tasks - p-tuning, adapter etc - that depend on dynamic seq len use global batch based sampling

We still use Apex for microbatch calculator, some enums/utils [both to be soon shifted to core],  LayerNorm/other things with kernel

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/build_model.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            set_defaults_if_not_set_tensor_model_parallel_attributes(param)
+
+    # Print number of parameters.
+    if parallel_state.model_parallel_is_initialized() and parallel_state.get_data_parallel_rank() == 0:
+        msg = " > number of parameters on (tensor, pipeline) model parallel rank ({}, {}): {}".format(
+            parallel_state.get_tensor_model_parallel_rank(),
+            parallel_state.get_pipeline_model_parallel_rank(),
+            _calc_number_of_params(model),
+        )
+        print(msg, flush=True)
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/build_model.py:142]: This should be a logging statement

Comment (aklife97) [reply nemo/collections/nlp/modules/common/megatron/build_model.py]: done!

**Review by ericharper — DISMISSED** [review]
LGTM. Thank you!

This PR was a massive effort. Thanks to all for their contributions and especially @aklife97 for putting it all together here.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**